File: built-in plugins dsp/mb-bell-chimes.cpp
/**
 * MB Wind Chimes
 * Category : instrument
 * Type     : bell
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Shimmering wind chimes with random tinkling
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BELL_CHIMES_H
#define MB_BELL_CHIMES_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBellChimes : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bell-chimes";
    static constexpr const char* PLUGIN_NAME    = "MB Wind Chimes";
    static constexpr const char* PLUGIN_TYPE    = "bell";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float density = 0.5f;  // range [0, 1]
    float brightness = 0.7f;  // range [0, 1]
    float damping = 0.4f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbBellChimes() = default;
    ~MbBellChimes() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.density = std::clamp(params.density, 0f, 1f);
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.damping = std::clamp(params.damping, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Wind Chimes
        return input;
    }
};

#endif // MB_BELL_CHIMES_H

File: built-in plugins dsp/mb-bell-chimes.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBellChimesPlugin: PluginDefinition = {
  id: "mb-bell-chimes",
  slug: "mb-bell-chimes",
  name: "MB Wind Chimes",
  category: "instrument",
  type: "bell" as any,
  version: "1.0.0",
  description: "Shimmering wind chimes with random tinkling",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.35 },
    { type: "sine", detune: 1500, gain: 0.25 },
    { type: "sine", detune: 2800, gain: 0.2 },
    { type: "sine", detune: 3900, gain: 0.2 },
  ],
  envelope: { attack: 0.001, decay: 1.5, sustain: 0.0, release: 1.0 },
  parameters: [
    {
      id: "density",
      name: "Density",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "damping",
      name: "Damping",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { density: 0.5, brightness: 0.7, damping: 0.4, volume: 0.7 },
};

export default MbBellChimesPlugin;
